polaris-react/src/utilities/use-index-resource-state.ts

```diff
@@ -41,12 +42,37 @@ export function useIndexResourceState<T extends {[key: string]: unknown}>(
     resourceFilter: undefined,
   },
 ) {
-  const [selectedResources, setSelectedResources] = useState(
-    initSelectedResources,
-  );
+  const [selectedResources, setSelectedResources] = useState<
+    ReadonlySet<string>
+  >(new Set(preCheckedResources));
   const [allResourcesSelected, setAllResourcesSelected] = useState(
     initAllResourcesSelected,
   );
+  const [unselectedResources, setUnselectedResources] = useState<
+    ReadonlySet<string>
+  >(new Set());
+  const [dirty, setDirty] = useState(false);
+
+  const prevPreCheckedResourcesRef = useRef(preCheckedResources);
+
+  useEffect(() => {
+    if (!isEqual(prevPreCheckedResourcesRef.current, preCheckedResources)) {
+      const filteredResources = dirty
+        ? preCheckedResources.filter(
+            (resource) => !unselectedResources.has(resource),
+          )
+        : preCheckedResources;
+
+      setSelectedResources(new Set(filteredResources));
+      prevPreCheckedResourcesRef.current = preCheckedResources;
+    }
+  }, [dirty, preCheckedResources, unselectedResources]);
```

Comment:
This makes sure that `preCheckedResources` are synced on pagination.

The current behavior does not re-hydrates on pagination since the state is never updated.